x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/SimplePattern.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.runtimefields.mapper;
+
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.LazyInitializable;
+import org.elasticsearch.dissect.DissectParser;
+import org.elasticsearch.grok.Grok;
+import org.elasticsearch.grok.MatcherWatchdog;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * A simple pattern that we can implement against grok and dissect.
+ */
+public interface SimplePattern {
```

Comment:
None of us is particularly happy with this name. Sure, right now it is like java's `Pattern` class but with only a single method so, sure, it is "simple" *now*. But "Simple" just sounds like a bad class name. `AbstractingPattern`? `NamedMatchesPattern`? `GrokOrDissectPattern`? `FlavoredPattern`? These sound kind of bad to me too. But maybe better than `SimplePattern`?

Reply:
If anyone can come up with a better name for this thing then I'm all ears.

Reply:
@rjernst Pinging for a name :)

Reply:
I wonder if we could use something other than Pattern, since this doesn't actually directly have anything to do with java regexes. In essence it is a functional interface to take a string, and return a map of named substrings. Maybe `NamedGroupExtractor`?

Reply:
`NamedGroupExtractor` will do just fine. We can always change it later because folks won't compile against it. I hope.